src/v/storage/translating_reader.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "model/fundamental.h"
+#include "model/record_batch_reader.h"
+#include "storage/offset_translator_data.h"
+
+#include <seastar/core/shared_ptr.hh>
+
+namespace storage {
+
+/// A struct that holds together a reader that performs offset translation and
+/// the corresponding offset translation data (useful when offsets of returned
+/// batches need to be translated back to the original offset space)
+struct translating_reader {
```

Comment:
can we make this `record_batch_reader::impl` ? This way we can use `record_batch_reader` abstraction

Reply:
Well, how would that help? This struct is just a glorified `std::pair` that is used to get access to `ot_data` in the fetch handler. It isn't very elegant but it is the best I could come up with :) So if we make this a `record_batch_reader::impl`, we can return `record_batch_reader ` from `partition_proxy::make_reader`, but would still need to somehow get the `ot_data` (a separate virtual method perhaps?)